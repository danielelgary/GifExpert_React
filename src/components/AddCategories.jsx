import { useState } from 'react'

export const AddCategories = ({ onNewCategory }) => {

    //Cada componente puede tener sus propios estados
    //manejara sus propios HOOKS para manejar el estado de los elementos propios
    const [inputValue, setInputValue] = useState('');
    const onInputChange = ({ target }) => {
        //El parametro de entrada {target} es la desestructuracion del evento
        //que se lanza desde el onChange (ChangeEventHandler) y el cual 
        //trae la info del valor que detecta que cambio.
        setInputValue(target.value);
    };

    const onSubmit = ( event ) => {
        //previene el full refresh de la pagina
        event.preventDefault();
        if (inputValue.trim().length < 1) return;

        //setCategories(categorias => [inputValue, ...categorias ]);
        onNewCategory( inputValue.trim() );
        setInputValue('');
    }


    return (
        //El comportamiento por defecto es que cuando se presiona enter 
        //haga un full refresh al formulario y trate de enviar los datos
        //para evitarlo, se usa el evento onSubmit 
        //y se maneja con un event.preventDefault
        <form onSubmit={onSubmit}>
            <input type={'text'} placeholder='Buscar Gif' value={inputValue} onChange={ onInputChange }/>

        </form>
    )
}