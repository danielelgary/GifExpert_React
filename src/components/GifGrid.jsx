import { GifItem } from "./index";
import { useFetchGifs } from "../hooks/useFetchGifs";



export const GifGrid = ({ category }) => {

    const {images, isLoading} = useFetchGifs( category );
    
    //Se comenta porque en esta clase se convirtio toda esta logica, en 
    //un hook personalizado que hara todo esto y dejara mas limpio el codigo.

    // const [images, setImages] = useState([])

    // const getImages = async() => {
    //     const newImages = await getGifs(category);
    //     setImages(newImages);
    // }

    // //El array vacio quiere decir que solo se dispara cuando se
    // //vuelve a cargar todo el componente.
    // useEffect(() => {    
    //     getImages();
    // }, [])
    
    return (
        <>
            <h3>{category}</h3>

            {
                //Forma 1:
                isLoading ? ( <h2>Cargando... </h2> ) : null
            }
            {
                //Forma 2:
                isLoading && ( <h2>Cargando... </h2> )
            }

            <div className="card-grid">
                {/* Se podria usar asi, sin embargo, se realiza con 
                desestructuracion para que quede mas presentable el codigo
                    {images.map( img => {
                        return <li key={img.id}>{img.title}</li>
                    })}
                */}
                {images.map( (img) => (
                    <GifItem 
                        key={img.id} 
                        { ...img } 
                        />
                    ))}
            </div>
        </>
    )
}
