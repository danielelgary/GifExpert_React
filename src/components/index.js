/*
Esto se conoce como un archivo de barril que sirve para exportar
todo lo que deseemos de manera global para ser usado de una manera mas sencilla
los imports y export en una aplicacion muy grande.

Al darle el * estamos diciendo que exporte todo lo que exporta 
el directorio que le se√±alemos.

Para importar en los componentes, ya solo seria importar el index.
*/

export * from './AddCategories'
export * from './GifGrid'
export * from './GifItem'