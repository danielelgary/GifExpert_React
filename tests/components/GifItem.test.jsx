import { GifItem } from "../../src/components/GifItem";
import { render, screen, fireEvent } from '@testing-library/react';

describe('Pruebas del archivo: <GifItem />', () => { 
    
    const title = 'Dragon Ball';
    const url = 'https://prueba.com/'

    test('Match con el Snapshot del componente', () => { 
        const { container } = render( <GifItem title={title} url={url} /> )

        expect( container ).toMatchSnapshot();
     });

     test('Debe mostrar imagen con el URL y ALT indicado', () => { 
        render( <GifItem title={title} url={url} /> );
        screen.debug();

        //Mirar todo lo que trae img o cualquier componente
        console.log(screen.getByRole('img'))

        //Se puede hacer de esta manera
        expect( screen.getByRole('img').src ).toBe(url);
        expect( screen.getByRole('img').alt ).toBe(title);

        //Pero si hay muchos, hacer desestructuracion
        const { src, alt } = screen.getByRole('img');

        expect(src).toBe(url);
        expect(alt).toBe(title);
      });

      test('Titulo se muestra en el componente', () => { 
        render( <GifItem title={title} url={url} /> );

        expect( screen.getByText(title) ).toBeTruthy()
        
       })
 });