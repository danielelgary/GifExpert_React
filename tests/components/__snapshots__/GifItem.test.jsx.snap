// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del archivo: <GifItem /> Match con el Snapshot del componente 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Dragon Ball"
      src="https://prueba.com/"
    />
    <p>
      Dragon Ball
    </p>
  </div>
</div>
`;
